### PR TITLE
GH-115802: JIT "small" code for macOS and Linux

### PR DESCRIPTION
**Pull request by brandtbucher**
Windows will come next, since that uses a different mechanism.

Somewhere in the neighborhood of [~0-3% faster](https://github.com/faster-cpython/benchmarking-public/blob/main/results/bm-20240221-3.13.0a4+-ff044c0-JIT/README.md), as expected.

* Issue: gh-115802

### CODE REVIEW

File: Python/jit.c [hunk 1]
```diff
@@ -249,10 +292,53 @@ patch(char *base, const Stencil *stencil, uint64_t *patches)
...
+                    (next_hole->kind == HoleKind_ARM64_RELOC_GOT_LOAD_PAGEOFF12 ||
+                     next_hole->kind == HoleKind_R_AARCH64_LD64_GOT_LO12_NC) &&
+                    next_hole->offset == hole->offset + 4 &&
+                    next_hole->symbol == hole->symbol &&
+                    next_hole->addend == hole->addend &&
+                    next_hole->value == hole->value)
+                {
+                    unsigned char rd = get_bits(loc32[0], 0, 5);
+                    assert(IS_AARCH64_LDR_OR_STR(loc32[1]));
+                    unsigned char rt = get_bits(loc32[1], 0, 5);
```
Comment (markshannon) [inline Python/jit.c:312]: `rt` and `rn` are only used in the assert, so I'm getting an unused variable warning.

File: Python/jit.c [hunk 2]
```diff
@@ -249,10 +292,53 @@ patch(char *base, const Stencil *stencil, uint64_t *patches)
...
+                    unsigned char rn = get_bits(loc32[1], 5, 5);
+                    assert(rd == rn && rn == rt);
```
Comment (markshannon) [inline Python/jit.c:314]:
```suggestion
                    assert(rd == get_bits(loc32[1], 5, 5) && rd == get_bits(loc32[1], 0, 5));
```

File: Tools/jit/_targets.py [hunk 3]
```diff
@@ -121,21 +122,14 @@ async def _compile(
             "-fno-builtin",
             # SET_FUNCTION_ATTRIBUTE on 32-bit Windows debug builds:
             "-fno-jump-tables",
-            # Position-independent code adds indirection to every load and jump:
-            "-fno-pic",
+            "-fno-plt",
```
Comment (markshannon) [inline Tools/jit/_targets.py:125]: Could you make sure each option has a comment saying what it does.

**Review by markshannon — CHANGES_REQUESTED** [review]
I minor issue with an assert, otherwise looks good.

**Comment by brandtbucher** [comment]
Sorry, looks like I merged just before your review. I'll address these in the follow-up Windows PR.

**Pull request merged**